### DISCUSSION TITLE
update button of the target file editor in eclipse updates to RC releases for mvn depedencies

### DISCUSSION THREAD

**Discussion by jcompagner** [eclipse-m2e/m2e-core#1584] [General]
I guess the implementation of the update button in the pde target file editor is done by m2e right?

Because if i push it i get a lot of updates to RC jars, which i don't want to have, i really want just final releases
and example is jackson annotations:

https://mvnrepository.com/artifact/com.fasterxml.jackson.core/jackson-annotations

if i press update now it will update to 2.16.0-RC1 but i want that it updates to 2.15.3 (i come from 2.15.2)

it seems that mvn knows its an RC (it has different colors) not sure if that is just in the version name or if it is something else.

**Comment by HannesWell** [comment]
> I guess the implementation of the update button in the pde target file editor is done by m2e right?

Correct.



> it seems that mvn knows its an RC (it has different colors) not sure if that is just in the version name or if it is something else.

As far as I know, the only strict rule for version names in Maven is that the `-SNAPSHOT` suffix means, well a snapshot. All other suffixes can be chosen more or less arbitrarily. Indeed most released artifacts have a `Major(.Minor(.Service))` version for their final releases but not all. Some use for example `-RELEASE` or `-FINAL` as suffix to distinguish final releases from milestones or release-candidates (I have not example at hand, but I have seen both). And for example Guava (https://mvnrepository.com/artifact/com.google.guava/guava) uses -jre or -android to distinguish different variants of the library.
So from a Maven point of view everything that is contained in repository like Maven-Central, which is not flagged as snapshot is an equally valid 'release'. Maven also knows if one Version is smaller, greater or equal than another but does not know anything for example about Semantic Versioning, which is just a recommended convention and not enforced in any way.
And at least for the case of Maven-Central RC or milestone artifacts are also retained forever just like 'final' releases. So such suffixes are more or less just for consumers to indicate that something might change in an unexpected manner in the next version (whatever it is).

I assume mvnrepository uses some kind of heuristic to estimate if a version suffix is a 'non-final' artifact, but since there are no strict rules such heuristics are always prone to false positive or negatives. And since one _can_ also use RC/M artifacts from M.-C. just like all other artifacts some event want to use them explicitly (of course one should be aware of what one is doing).

**Comment by laeubi** [comment]
Maven has special handling for M / RC / A and alike what affects ordering, m2e simply performs an lookup for the latest version as it works in most cases well and is easy to implement.

You can see this in code [here](https://github.com/apache/maven-resolver/blob/12b1c2cb27106c647c6bd8a1254e40d4b0530682/maven-resolver-util/src/main/java/org/eclipse/aether/util/version/GenericVersion.java#L171).

**Comment by jcompagner** [comment]
i also tested a bit the mvn versions plugin, but you really need todo this:

```
<plugin>
    <groupId>org.codehaus.mojo</groupId>
    <artifactId>versions-maven-plugin</artifactId>
    <version>2.16.1</version>
    <configuration>
        <ignoredVersions>.*-rc.*,.*-alpha.*</ignoredVersions>
    </configuration>
</plugin>
```

to remove the "rc" or "alpha" stuff from it
so maven doesn't really have that distinction i guess build like (like npm has with tags)

so i guess that is just pressing that button and then in the compare window go one by one over them to check the version if it makes sense is the only real way todo this.
